### PR TITLE
Matthew's Homework

### PR DESCRIPTION
**Pull request by matthewdeepblue**

### CODE REVIEW

**Comment by rfk** [comment]
Thanks @matthewdeepblue! I note that this is a public pull request, so I'm going to close it in order to make it less visible to other potential candidates, in case anyone wants to copy your homework 😅

File: README_Matthew.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+terraform init
+terraform import module.s3_bucket.aws_s3_bucket.this[0] annalise-ai-datalake
+terraform plan
+
+- Describe how you would copy the data in `annalise-ai-datalake` to `harrison-ai-landing`.  You may wish to provide Terraform templates for any resources that would be used by this process, but it is not required.
+Create s3 access point and IAM Role on harrison-ai-landing
+Create VPCE and IAM Role on annalise-ai-datalake
+Share harrison-ai-landing kms key with annalise-ai account
+aws sts assume-role —role-are "harrison-ai-landing IAM Role" —role-session-name "Data-copy"
+aws s3 cp s3://SOURCE_BUCKET_NAME/pathxxx/* s3://"s3 access point alias"/pathxxx/* —sse aws:kms —sse-kms-key-id "harrison-ai-landing kms arn"
```
Comment (rfk) [inline README_Matthew.md:20]: Can you suggest any ways to improve the performance and/or reliability of this step when the bucket contains a large number of files?

File: list_s3_objects.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+aws_secret_access_key = 'YOUR_SECRET_ACCESS_KEY'
+region_name = 'ap-southeast-2'
+bucket_name = 'YOUR_BUCKET_NAME'
+queue_url = 'YOUR_QUEUE_URL'
+session = boto3.Session()
+s3_client = session.client('s3', aws_access_key_id=aws_access_key_id, aws_secret_access_key=aws_secret_access_key, region_name=region_name)
+sqs_client = session.client('sqs', aws_access_key_id=aws_access_key_id, aws_secret_access_key=aws_secret_access_key, region_name=region_name)
+
+def list_s3_objects(bucket):
+    response = s3_client.list_objects(Bucket=bucket_name)
```
Comment (rfk) [inline list_s3_objects.py:16]: How would you expand this to deal with a bucket containing more than 1000 objects, thus requiring pagination?

**Review by rfk — COMMENTED** [review]
Hi @matthewdeepblue, thanks for diving in here! I just had a couple of follow-up questions based on our review session for your tech task, see below.

File: README_Matthew.md (hunk 1 above)
Comment (matthewdeepblue) [reply README_Matthew.md:20]: There are multiple ways which we could use to transfer a large number of files between S3 buckets.
AWS datasync,  we could set both source and destination to s3 and also use CloudWatch to monitor the transfer.
AWS S3 batch replication,  we could create a replication rule to copy data and also generate a report.
AWS CLI,  we could use a robust ec2 server to cp data in parallel , data could be split by prefix or folder.

File: list_s3_objects.py (hunk 2 above)
Comment (matthewdeepblue) [reply list_s3_objects.py:16]: Pagination is used in boto3 client.  To parallelize the listing process, we could use asyncio and ProcessPoolExecutor to deal large number of objects.

**Pull request closed without merging**